src/optimizer/simple_optimizer.cpp

```diff
@@ -1508,5 +1363,164 @@ bool SimpleOptimizer::UnderlyingSameOrder(planner::AbstractPlan* select_plan,
   LOG_TRACE("All checking is done, so ordering is the same");
   return true;
 }
+
+std::unique_ptr<planner::AbstractPlan> SimpleOptimizer::CreateOrderByPlan(
+    parser::SelectStatement* select_stmt, planner::AbstractPlan* child_plan,
+    catalog::Schema* schema, std::vector<oid_t> column_ids, bool is_star) {
+  auto schema_columns = schema->GetColumns();
+  std::vector<oid_t> keys;
+  // The order_by plan is oblivious to the type of query, so,
+  // the columns must be manually added for it.
+  if (is_star) {
+    for (size_t column_ctr = 0; column_ctr < schema_columns.size();
+         column_ctr++) {
+      keys.push_back(column_ctr);
+    }
+  } else {
+    for (size_t column_ctr = 0; column_ctr < select_stmt->select_list->size();
+         column_ctr++) {
+      keys.push_back(column_ctr);
+    }
+  }
+  std::vector<bool> flags;
+  if (select_stmt->order->type == 0) {
+    flags.push_back(false);
+  } else {
+    flags.push_back(true);
+  }
+  std::vector<oid_t> key;
+  std::string sort_col_name(
+      ((expression::TupleValueExpression*)select_stmt->order->expr)
+          ->GetColumnName());
+  for (size_t column_ctr = 0; column_ctr < schema_columns.size();
+       column_ctr++) {
+    std::string col_name(schema_columns.at(column_ctr).GetName());
+    if (col_name == sort_col_name) {
+      if (is_star) {
+        // The whole schema is already added, and column_ids
+        // doesn't represent faithfully the columns.
+        key.push_back(column_ctr);
+      } else {
+        // The column_ctr is not reliable anymore given that we are
+        // looking to the whole schema.
+        // Since the columns were added in the column_ids, it is safe
+        // to retrieve only those indexes.
+        auto iter = find(column_ids.begin(), column_ids.end(), column_ctr);
+        auto column_offset = std::distance(column_ids.begin(), iter);
+        key.push_back(column_offset);
+      }
+    }
+  }
+
+  std::unique_ptr<planner::OrderByPlan> order_by_plan(
+      new planner::OrderByPlan(key, flags, keys));
+
+  // Whether underlying child's output has the same order with the
+  // order_by clause.
+  LOG_TRACE("order by column id is %d", schema->GetColumnID(sort_col_name));
+  if (UnderlyingSameOrder(child_plan, schema->GetColumnID(sort_col_name),
+                          flags.front()) == true) {
+    LOG_TRACE(
+        "Underlying plan has the same ordering output with"
+        "order_by plan without limit");
+    order_by_plan->SetUnderlyingOrder(true);
+  }
+  return std::move(order_by_plan);
+}
+
+std::unique_ptr<planner::AbstractPlan> SimpleOptimizer::CreateOrderByLimitPlan(
+    parser::SelectStatement* select_stmt, planner::AbstractPlan* child_plan,
+    catalog::Schema* schema, std::vector<oid_t> column_ids, bool is_star) {
```

Comment:
Got it, I tried not to mess around with the already present code, but I'll do that.